examples/hacspec-gimli/src/lib.rs

```diff
@@ -0,0 +1,251 @@
+            // Absorb full blocks
+            let full_block = Block::from_seq(&input_block);
+            s = absorb_block(full_block, s);
+        } else {
+            // Absorb last incomplete block
+            // Note that this would work in all other cases as well, but the above is safer.
+            let input_block_padded = Block::new();
+            let mut input_block_padded = input_block_padded.update_start(&input_block);
+            input_block_padded[block_len] = U8(1u8);
+
+            // XOR in capacity part
+            s[11] = s[11] ^ U32(0x01000000u32);
+            s = absorb_block(input_block_padded, s);
+        }
+    }
+
+    s
+}
+
+pub fn gimli_hash(input_bytes: &ByteSeq) -> Digest {
+    let s = State::new();
+    let s = gimli_hash_state(input_bytes, s);
+    let output = Digest::new();
+    let output = output.update_start(&squeeze_block(s));
+    let s = gimli(s);
+    output.update(Block::length(), &squeeze_block(s))
+}
+
+// === Cipher ===
+
+bytes!(Nonce, 16);
+bytes!(Key, 32);
+public_bytes!(Tag, 16);
+
+fn process_ad(ad: &ByteSeq, s: State) -> State {
+    gimli_hash_state(ad, s)
+}
+
+fn process_msg(message: &ByteSeq, mut s: State) -> (State, ByteSeq) {
+    let mut ciphertext = ByteSeq::new(message.len());
+
+    let rate = Block::length();
+    let num_chunks = message.num_chunks(rate);
+    for i in 0..num_chunks {
+        let key_block = squeeze_block(s);
+
+        let (block_len, msg_block) = message.get_chunk(rate, i);
+        // This pads the msg_block if necessary.
+        let msg_block_padded = Block::new();
+        let mut msg_block_padded = msg_block_padded.update_start(&msg_block);
+        ciphertext = ciphertext.set_chunk(
+            rate,
+            i,
+            // the slice_range cuts off the last block if it is padded
+            &(msg_block_padded ^ key_block).slice_range(0..block_len),
+        );
+        if i == num_chunks - 1 {
+            // XXX: this is in the code but I don't see it in the spec.
+            msg_block_padded[block_len] = msg_block_padded[block_len] ^ U8(1u8);
+            s[11] = s[11] ^ U32(0x01000000u32); // s_2,3
+        }
+        s = absorb_block(msg_block_padded, s);
+    }
+
+    (s, ciphertext)
+}
+
+fn process_ct(ciphertext: &ByteSeq, mut s: State) -> (State, ByteSeq) {
+    let mut message = ByteSeq::new(ciphertext.len());
+
+    let rate = Block::length();
+    let num_chunks = ciphertext.num_chunks(rate);
+    for i in 0..num_chunks {
+        let key_block = squeeze_block(s);
+        let (block_len, ct_block) = ciphertext.get_chunk(rate, i);
+
+        // This pads the ct_block if necessary.
+        let ct_block_padded = Block::new();
+        let ct_block_padded = ct_block_padded.update_start(&ct_block);
+        let msg_block = ct_block_padded ^ key_block;
+
+        // XXX: pretty ugly. Zero pad the last block if it is padded.
+        let mut msg_block = Block::from_slice_range(&msg_block, 0..block_len);
```

Comment:
You can move that down to after line 166 where you set the new chunk no ?

Reply:
This actually got messed up a little. I cleaned it up.